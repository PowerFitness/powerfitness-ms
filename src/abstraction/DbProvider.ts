import { DbConfig } from '../config'
import { createConnection, Connection, MysqlError } from 'mysql';

export class DbProvider {
	dbConfig: DbConfig;
	connection: Connection | null = null;
	constructor(dbConfig: DbConfig) {
		this.dbConfig = dbConfig;
	}
	establishConnection(): void {
		if(!this.connection) {
			this.connection = createConnection(this.dbConfig)
		}
		return;
	}
	execute<T>(query: string): Promise<T> {
		this.establishConnection();
		return new Promise((resolve, reject) => {
			(this.connection as Connection).query(query, (error: MysqlError, results: T) => {
				if (error) {
					reject(error)
				} else {
					resolve(results)
				}
				(this.connection as Connection).end()
			})
		}) as unknown as Promise<T>;
	}

	beginTransaction(): Promise<void> {
		this.establishConnection();
		return new Promise((resolve, reject) =>
			(this.connection as Connection).beginTransaction((err: MysqlError) => err ? reject(err) : resolve()));
	}

	commit(): Promise<void> {
		this.establishConnection();
		return new Promise((resolve, reject) => (this.connection as Connection).commit((err: MysqlError) => {
			if (err) {
				this.rollback()
					.then(() => reject(err))
					.catch((rollbackError: MysqlError) =>
						reject(new Error(`Rollback Failed: ${err.message} / ${rollbackError.message}`)))
					.finally(() =>
						(this.connection as Connection).end())
			} else {
				resolve();
				(this.connection as Connection).end()
			}
		}));
	}

	rollback(): Promise<void> {
		return new Promise((resolve, reject) =>
			(this.connection as Connection).rollback((err: MysqlError) => err ? reject(err) : resolve()));
	}
}

export default DbProvider;
