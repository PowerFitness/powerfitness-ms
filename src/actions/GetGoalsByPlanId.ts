import { Action } from '../abstraction/Action';
import DbProvider from '../abstraction/DbProvider';
import { escape } from 'mysql';

export interface Goal {
    id: number;
    planId: number;
    type: string;
    name: string;
    unit: string;
    value: number;
    createdDate: string;
    updatedDate: string;
}

export class GetGoalsByPlanId extends Action<Array<Goal>> {
	planId: string;
	constructor(dbProvider: DbProvider, planId: string) {
		super(dbProvider);
		this.planId = planId;
	}
	buildQuery(): string {
		if (!this.planId) throw new Error('Missing criteria');
		return `select * from goal where planId = ${escape(this.planId)}`;
	}
}

export default GetGoalsByPlanId;
