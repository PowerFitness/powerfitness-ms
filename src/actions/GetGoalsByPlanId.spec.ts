import DbProvider from '../abstraction/DbProvider';
import { mock } from 'jest-mock-extended';
import GetGoalsByPlanId from './GetGoalsByPlanId';

describe('GetGoalsByPlanId', () => {
	test('buildQuery', () => {
		const dbProvider: DbProvider = mock<DbProvider>();
		const action: GetGoalsByPlanId = new GetGoalsByPlanId(dbProvider, '1');
		expect(action.buildQuery()).toMatchSnapshot();
	})
	test('buildQuery throws an error if no criteria', () => {
		const dbProvider: DbProvider = mock<DbProvider>();
		const action: GetGoalsByPlanId = new GetGoalsByPlanId(dbProvider, undefined as unknown as string);
		expect(action.buildQuery).toThrowError()
	})
})
