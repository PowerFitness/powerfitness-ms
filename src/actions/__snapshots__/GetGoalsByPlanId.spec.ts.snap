// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GetGoalsByPlanId buildQuery 1`] = `"select * from goal where planId = '1'"`;
